design/ceph/multus-network.md

```diff
@@ -78,7 +78,123 @@ Nothing to do in particular since they don't use any service IPs.
+not become blocked. The Rook-Ceph Operator's CSI controller should set the DaemonSet's update
+strategy to `OnDelete` so that the pods do not get deleted if the DaemonSet is updated while also
+ensuring that the pods will be updated on the next node reboot (or node drain).
+
+#### Mover DaemonSet and Pods
+The Rook-Ceph Operator's CSI controller also creates a second DaemonSet configured to use host
+networking. This DaemonSet also runs on all nodes that will have CSI plugin pods (and holder pods).
+Mover pods exist to "move" the multus network interface being held by the holder pod on the node
+into the host's network namespace to provide user's volumes with uninterrupted access to the Ceph
+cluster, even when the CSI driver is restarted (or updated).
+
+The mover must:
+- be a privileged container
+- have `SYS_ADMIN` and `NET_ADMIN` capabilities
+- be on the host network namespace
+- have access to the `/var/run/netns` directory
+
+In order to not leave moved interfaces dangling in the host's network namespace, mover pods must
+move interfaces back to their original namespace when CSI is being terminated. The most
+straightforward way to accomplish this is to move interfaces back when the mover is being
+terminated. If an interface is moved without user applications also being removed, this will cause
+I/O disruption. Therefore, the DaemonSet should also use the `OnDelete` update strategy so that the
+pods can be updated on node reboots (or node drains).
+
+In order to better handle unexpected corner cases that leave moved interfaces in the host network
+namespace (e.g., a mover is killed abruptly rather than gracefully terminated), instead treat "move"
+operations as a disable-and-copy operation. To do this, disable the interface in the holder pod's
+network namespace, and create a copy of the interface in the host namespace with the same MAC
+address and IP config. From the user standpoint, the interface is still "moved" because the original
+is disabled, so we keep the "mover" terminology. This merely helps Rook ensure that it is not
+accidentally losing the original information.
+
+A previous iteration of this design specified the mover application as a sidecar to Ceph CSI plugin
+pods; however, this design would mean that the mover would need to be deleted and re-created
+whenever the CSI plugin is updated, possibly resulting in I/O hangs in user pods during the update.
+Keeping the mover independent allows CSI plugin updates to happen freely without complex
+interactions between it and the mover.
+
+#### Interactions between components
+If a copied interface is left in the host network namespace after the holder pod is removed, multus
+may later give the address to a different application, and the CSI driver may try to connect to the
+different application with Ceph requests. We should try to avoid leaving interfaces on the host as
+much as possible. Killing the mover pod abruptly will leave copied interfaces, but there is no way
```

Comment:
There is a fine line here, we can decide to handle and ignore SIGTERM, if it is sent from the host. However, if it's pod termination (through k8s API), once the grace period is over the process will receive SIGKILL which cannot be ignored.

Reply:
I think we can't ignore SIGTERM if we want to clean up when the mover stops. Related to the race condition I mention here: https://github.com/rook/rook/pull/9643#discussion_r799701722